elasticsearch5/search-index/src/main/java/org/vertexium/elasticsearch5/bulk/BulkUpdateService.java

```diff
@@ -216,21 +226,99 @@ private void complete(BulkItem bulkItem, Exception exception) {
-            } else {
-                throw new VertexiumException("unhandled request type: " + actionRequest.getClass().getName());
+
+        Map<String, List<BulkItem>> byIndexName = bulkItems.stream().collect(Collectors.groupingBy(BulkItem::getIndexName));
+        for (Map.Entry<String, List<BulkItem>> byIndexNameEntry : byIndexName.entrySet()) {
+            String indexName = byIndexNameEntry.getKey();
+            Map<String, List<BulkItem>> byType = byIndexNameEntry.getValue().stream().collect(Collectors.groupingBy(BulkItem::getType));
+
+            for (Map.Entry<String, List<BulkItem>> byTypeEntry : byType.entrySet()) {
+                String type = byTypeEntry.getKey();
+                Map<String, List<BulkItem>> byDocumentId = byTypeEntry.getValue().stream().collect(Collectors.groupingBy(BulkItem::getDocumentId));
+
+                for (Map.Entry<String, List<BulkItem>> byDocumentIdEntry : byDocumentId.entrySet()) {
+                    String documentId = byDocumentIdEntry.getKey();
+                    List<BulkItem> requestBulkItems = byDocumentIdEntry.getValue();
+                    Map<Class<? extends BulkItem>, List<BulkItem>> byBulkItemType = requestBulkItems.stream().collect(Collectors.groupingBy(BulkItem::getClass));
+                    if (byBulkItemType.size() > 1) {
+                        throw new VertexiumException(
+                            "Cannot mix bulk item types in a batch: "
+                                + byBulkItemType.keySet().stream().map(Class::getName).collect(Collectors.joining(","))
+                        );
+                    }
+                    for (Map.Entry<Class<? extends BulkItem>, List<BulkItem>> byBulkItemTypeEntry : byBulkItemType.entrySet()) {
+                        Class<? extends BulkItem> bulkItemType = byBulkItemTypeEntry.getKey();
+                        if (bulkItemType == UpdateBulkItem.class) {
+                            Map<String, String> source = new HashMap<>();
+                            Map<String, Object> fieldsToSet = new HashMap<>();
+                            List<String> fieldsToRemove = new ArrayList<>();
+                            Map<String, String> fieldsToRename = new HashMap<>();
+                            List<String> additionalVisibilities = new ArrayList<>();
+                            List<String> additionalVisibilitiesToDelete = new ArrayList<>();
+
+                            boolean updateOnly = true;
+                            for (BulkItem bulkItem : requestBulkItems) {
+                                UpdateBulkItem updateBulkItem = (UpdateBulkItem) bulkItem;
+                                source.putAll(updateBulkItem.getSource());
+                                fieldsToSet.putAll(updateBulkItem.getFieldsToSet());
+                                fieldsToRemove.addAll(updateBulkItem.getFieldsToRemove());
+                                fieldsToRename.putAll(updateBulkItem.getFieldsToRename());
```

Comment:
It might make debugging a little easier if you throw an error on collisions here. If you have two BulkItems renaming such as:
1. `a` -> `b`
1. `a` -> `b`
That should be OK. But something like:
1. `a` -> `b`
1. `a` -> `c`
Should raise an error. 

I think it might blow up in the Painless script anyway, but errors in Painless are really hard to figure out.